core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesClient.scala

```diff
@@ -0,0 +1,218 @@
+        val pod =
+          kubeRestClient.pods().withName(id.asString).waitUntilReady(timeouts.inspect.length, timeouts.inspect.unit)
+        ContainerAddress(pod.getStatus().getPodIP())
+      }
+    }.recoverWith {
+      case e =>
+        log.error(this, s"Failed to get IP of Pod '${id.asString}' within timeout: ${e.getClass} - ${e.getMessage}")
+        Future.failed(new Exception(s"Failed to get IP of Pod '${id.asString}'"))
+    }
+  }
+
+  def rm(id: ContainerId)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", id.asString), timeouts.rm).map(_ => ())
+
+  def rm(key: String, value: String)(implicit transid: TransactionId): Future[Unit] =
+    runCmd(Seq("delete", "--now", "pod", "-l", s"$key=$value"), timeouts.rm).map(_ => ())
+
+  def logs(id: ContainerId, sinceTime: Option[String])(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    Source.fromFuture(Future {
+      blocking {
+        val path = Path / "api" / "v1" / "namespaces" / kubeRestClient.getNamespace / "pods" / id.asString / "log"
+        val query = Seq("timestamps" -> true.toString) ++ sinceTime.map(time => "sinceTime" -> time)
+        val url = Uri(kubeRestClient.getMasterUrl.toString)
+          .withPath(path)
+          .withQuery(Query(query: _*))
+
+        val request = new Request.Builder().get().url(url.toString).build
+        val response = kubeRestClient.getHttpClient.newCall(request).execute
+        if (response.isSuccessful) {
+          Future.successful(response.body.string)
+        } else {
+          Future.failed(
+            new Exception(s"Kubernetes API returned HTTP status ${response.code} when trying to retrieve pod logs"))
+        }
+      }
+    }.flatMap(identity).map { output =>
+      // The k8s logs api has less granularity than the timestamp
+      // we're passing as the 'sinceTime' argument, so we may need to
+      // filter logs from previous activations on a warm runtime pod
+      val original = output.lines.toSeq
+      val relevant = original.dropWhile(s => !s.startsWith(sinceTime.getOrElse("")))
+      val result =
+        if (!relevant.isEmpty && original.size > relevant.size) {
+          // drop matching timestamp from previous activation
+          relevant.drop(1)
+        } else {
+          original
+        }
+
+      // map the logs to the docker json file format expected by
+      // ActionLogDriver.processJsonDriverLogContents
+      ByteString(
+        result
+          .map { line =>
+            val pos = line.indexOf(" ")
+            val ts = line.substring(0, pos)
+            val msg = line.substring(pos + 1)
+            // TODO - when we can distinguish stderr: https://github.com/kubernetes/kubernetes/issues/28167
+            val stream = "stdout"
+            LogLine(ts, stream, msg).toJson.compactPrint
+          }
+          .mkString("\n") + "\n" // trailing newline is necessary or the frame won't be decoded and break the akka stream
```

Comment:
Why not just add the `"\n"` on the line above:

```scala
LogLine(ts, stream, msg).toJson.compactPrint + "\n"
```

Reply:
Makes sense to me, I think. :)

Reply:
the "why" of the trailing newline is my fault; no logic to it and the suggested adjustment makes sense. 

If I'm understanding you correctly, @markusthoemmes… right now we're splitting the entire logging result based on the HTTP response, and pushing it out as a single item. Your suggestion is to break it out so the stream takes individual events for each line, yes?

Reply:
@bwmcadams Yep, it goes through a framing stage immediatley anyway so I don't expect much difference in either way. It just feels clearer to have it immediately exploded into the specific events and squash them into the stream instead of creating a string just to split it immediatly.

Reply:
Actually, thinking about it: If you choose to implement my proposal you can just drop the FramingStage below.